server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java

```diff
@@ -85,9 +110,83 @@ private FieldFetcher(List<FieldContext> fieldContexts) {
                 documentFields.put(field, new DocumentField(field, parsedValues));
             }
         }
+        collectUnmapped(documentFields, sourceLookup.loadSourceIfNeeded(), "", 0);
         return documentFields;
     }
 
+    private void collectUnmapped(Map<String, DocumentField> documentFields, Map<String, Object> source, String parentPath, int lastState) {
+        for (String key : source.keySet()) {
+            Object value = source.get(key);
+            String currentPath = parentPath + key;
+            int currentState = step(this.unmappedFetchAutomaton, key, lastState);
+            if (currentState == -1) {
+                // current path doesn't match any fields pattern
+                continue;
+            }
+            if (value instanceof Map) {
+                // one step deeper into source tree
+                collectUnmapped(
+                    documentFields,
+                    (Map<String, Object>) value,
+                    currentPath + ".",
+                    step(this.unmappedFetchAutomaton, ".", currentState)
+                );
+            } else if (value instanceof List) {
+                // iterate through list values
+                List<Object> list = collectUnmappedList(documentFields, (List<?>) value, currentPath, currentState);
```

Comment:
It took me a while to see how this could be an issue but I think you mean cases like:
```
{
	"user": [{
		"name": "eve",
		"scores": [1, 2, 3]
	}, {
		"name": "adam",
		"scores": [4, 5, 6]
	}]
}
```
I see the `fields` API currently returns `"user.scores": [ 1, 2, 3, 4, 5, 6]` here at the moment, so I will add a test to make the unmapped behaviour doing the same. Or did you think about a different case?

Reply:
Interesting variation that cannot happen with mapped fields, but is accepted e.g. with `"dynamic": false`, since we have no mapping and type checks then:
```
{
	"user": [{
		"name": "eve",
		"scores": [1, 2, 3]
	}, {
		"name": "adam",
		"scores": "foo"
	}]
}
```
I think this should return the flattened `"user.scores": [ 1, 2, "foo"]` then, but maybe you disagree on that?

Reply:
Sorry for the confusion -- I didn't spot a behavioral bug, I just found the API choice surprising. The `collectUnmappedList` method returns a list, but `collectUnmapped` adds directly to the document fields map. I was wondering if there needs to be a difference?

Reply:
It looks like this comment is still open. Concretely, could we avoid returning a list from `collectUnmappedList` and just have it add to `documentFields`? This would match the signature for `collectUnmapped`.

Reply:
Thanks, I understand now. I investigated this and unfortunately it is not as straight-forward as it seems, at least I didn't find an easy way after a bit of searching. I now remember the reason I added the return type was another case I have in the tests. Consider this:
```
Mapping:
{
	"_doc": {
		"properties": {
			"obj": {
				"type": "object",
				"enabled": "false"
			}
		}
	}
}

Doc:
{
	"obj": ["string_value", {
			"a": "b"
		},
		[1, 2, 3]
	]
}
```
I would expect the fields output for `obj` to be `[ "string_value, [1, 2, 3]]` (note the list inside the list. Do do this I need to be able to add to the list collected when on the "obj" level (see  https://github.com/elastic/elasticsearch/pull/65386/files#diff-3c196c8a285307ef760ada7bc5a1df27dce1f4e7160bf75d028833793d171a9aR185). 
At the same time the case above (adding to the same key from two different nested objects should return a single concatenated list) should work. I wasn't able to reconcile both cases without the return type so far.
If you agree with the above expectations on list-inside-list maybe we can leave this investigation for a follow up as well since its an internal anyway?

Reply:
Got it, that is indeed a tricky example. I would actually expect that lists of lists get flattened, as we do for mapped fields. For example, we accept and index documents like this (although maybe we shouldn't):

```
{ "field": [[1, 2, 3]] }
```

The `fields` option will descend into the arrays and return `{ "field": [1, 2, 3] }`. So maybe we can just do the same in your example and return `{ "field": ["string_value", 1, 2, 3] }`.

Reply:
Got it, with those expectations I think I can alter the code and test to get around without the return type. I'm not particularly interested in the expected behaviour I mentioned earlier as long as you think flattening is okay here, and since that seems to be the behaviour for mapped types as well I think its okay.

Reply:
See latest commit and let me know if you are okay with this, otherwise I'd merge and backport to 7.x once CI is green